### ISSUE TITLE
Create a badging engine

### ISSUE THREAD

**Issue by prjseal** [prjseal/Umbraco-Community-Quiz#14]
Currently the badges are awarded in the code on submit of a form. It would be good to create some sort of badging engine. Perhaps it can raise an event and we hook into those events, check for the badge types for that type of event and then check if the data associated with this event matches the criteria for the badge and then award the badge.

**Comment by cornehoskam** [comment]
That sounds like a good addition! There would be many ways to setup such a system, but a more event driven notification & notificationHandler based system with a Badge Service to process the various notifications could make it quite flexible! 🤔

**Comment by prjseal** [comment]
Yes definitely. We want to raise event when things happen. Then hook into those notifications to consult with the badge engine to see if a badge is required and then apply it

**Comment by prjseal** [comment]
Yeah sure. Go for it. Please can you split it into separate PRs though. 1 for raising notifications throughout the site. Then one that consumes those notifications and works out which badges to give if any

I'd like to manage the rules of what badges to give for what using Umbraco

**Comment by cornehoskam** [comment]
Sure thing! I'll start with raising notificaitons at the locations the application currently hands out badges, and ones that one is approved, we can get working on process those notifications to hand out badges and removing the old badging code in the process!

**Comment by prjseal** [comment]
Brilliant. Thank you so much for joining in with this and helping make it a reality

**Comment by prjseal** [comment]
@cornehoskam please can you add an event for member logged in

**Comment by cornehoskam** [comment]
Sure thing @prjseal, will work on that next!

**Comment by prjseal** [comment]
Excellent, would you mind creating events for failed registration, failed edit profile and failed submit survey and failed submit question? If you have time?

**Comment by cornehoskam** [comment]
Sure! By Survey, do you mean a Quiz, or is a survey a new feature you have in mind?

**Comment by prjseal** [comment]
Yeah I meant quiz. Don't know why I said survey 🤣

**Comment by cornehoskam** [comment]
There you go Paul! I also took the liberty to add a notification for Failed login too, to stay consistent and have a failed counterpart for all the current Notifications!

**Comment by prjseal** [comment]
Brilliant. Thank you. I'll review it later

**Comment by cornehoskam** [comment]
Lovely! I'll do some thinking about a potential Badging Engine setup now that we have these notifications in place! I'd love to hear if you already have some idea's in mind @prjseal, otherwise I'll let my creativity go lose haha. I'll make sure to share some idea's with you before starting on any code either way 🙂

**Comment by cornehoskam** [comment]
Hi @prjseal, I've created a basic design of how I thought the assigning of badges using the new notifications could work! The essence would be that there is a Badge implementation for each badge in Umbraco, on which the logic for the condition resides. If that condition is met, the various notificationhandlers pass said badge to the badging service to assign the badge to the corresponding users! This would fully decouple the assigning of badges from any controllers, and have NotificationHandlers do the job for us, without having the handlers store the badging logic themselves! I'd love to hear your opinion on this before writing any code 🙂

![CommunityQuiz](https://user-images.githubusercontent.com/6321978/196615385-5094c21d-c1f1-4946-a663-99bab31e0aa3.jpg)

**Comment by cornehoskam** [comment]
We could solve that with the alias property on a c# badge, which could be the unique alias as a constant/static! We could even reference the ModelsBuilder model from there if we want, but I think using the alias would make it a bit more generic.

**Comment by prjseal** [comment]
Yeah I like that idea. This way we can create as many badges as we want in the content and then when it comes to implementing the logic we can just look for the alias

**Comment by cornehoskam** [comment]
Because the logic of the badge can be very context dependand, and we only provide the nessecary context through the notification itself (and we don't want make the IBadge implementation dependand on the notification), I opted to keep the logic for the badge condition in the notificationhandler for said specific badge to be the most logical place

**Comment by prjseal** [comment]
This is now live and is very exciting. Thank you so much. Fixed in #48
